SDK governance draft (#649)
cloudevents/spec: governance/SDK.md

## SDK Community

The community is organized as follows:

- Every SDK has its own
  [Github Team](https://github.com/orgs/cloudevents/teams), for example
  `sdk-go-maintainers` is the group of mantainers for
  [cloudevents/sdk-go project](https://github.com/cloudevents/sdk-go)
- The union of all the `sdk-*-maintainers` assembles the _sdk maintainers_ group

## New maintainers

TBD

## Ensuring projects health

In order to ensure the health of the SDK projects, we define the following
actions. These are crucial to show the commitment of the community to develop
and support these libraries, which are an important part of the CloudEvents
ecosystem.

We define an SDK project _healthy_ if:

1. It works with the latest version of the programming language
2. It supports the latest versions of the integrated libraries/frameworks
3. It receives security patches regularly
4. It supports the last N-1 major versions of CloudEvents spec, as defined in
   [Contribution Acceptance](../SDK.md#contribution-acceptance)
5. Issues and PRs are triaged (labeled, commented, reviewed, etc) regularly

> We define a project `cloudevents/sdk-x` _not actively maintained_ if:

Reviewer: What is an SDK has no activity because it doesn't need any work done on it? No language updates, no CE updates, no issues, etc... it's just working.

Author: Then It's fine, I don't define any actions in such cases. I define actions when issues/PR are not being triaged (archive/handover the project) or if the security patches are not applied and released.

Author: Maybe should i remove the criteria 6?

Reviewer: Yep, number 6 is the one that got me thinking about this... I agree we need something, just not sure how to word it.
I wonder if instead of being too prescriptive, it should be a bit more vague to allow for the other sdk maintainers to use their best judgement.  Something like:
```
There has been no necessary activity within the SDK for at least 4 months. The other SDK maintainers will  evaluate what "no necessary activity" means, but often this could mean "no commits", or "no issue
discussions". However, if the SDK is stable and does not need to be update then it might be determined
that it current state is acceptable.
```
I don't know...

Author: I prefer to keep it as note more than as criteria, because it's ok to have it vague but it's too much open to "interpretations" to be a criteria. let me try to reword it

> 6. No commits on master from `sdk-x-maintainers` since 4 months

Reviewer: s/since/for/

> 7. Issues and/or PRs are not being triaged from `sdk-x-mantainers` since 2

Reviewer: s/since/for/

   months

> 8. Security patches are not being **released** from `sdk-x-mantainers` since 1

Reviewer: s/since/for/

### Security patches

If a project `cloudevents/sdk-x` meets _healthy_ criteria 1, 2, 4, but the
`sdk-x-maintainers` group is not actively performing security patches as defined
in criteria 8, the community could decide to entitle one or more members from
the larger _sdk maintainers_ group to perform security patches and release new
minor versions of the project.

Because the people entitled to perform security patches need write permissions on the repository, the community MUST temporarily grant these permissions.
After the release of the patch, these permissions MUST be revoked.

Because of the urgent nature of security patches, an _informal_ vote during the
CloudEvents SDK meeting is enough to assign the security patch to the _sdk
maintainer(s)_ who volunteers to take care of it.

NOTE: this process is valid only for security patches, not for bugs or
enhancements.

### Handover to a new maintainer/group of maintainers

If a project is not meeting the criteria 1, 2, 4, 5, 6, 7, 8, the community
MAY decides to handover the project to a new maintainer/group of
maintainers. The community can perform the handover to a new maintainer
if all the following conditions are met:

> * A vote to add the new maintainer MUST have **already taken place** with

Reviewer: Who initiates a vote if all maintainers have vanished?

Reviewer: Perhaps the larger SDK maintainers should do it?

Reviewer: oh, I guess it's implied that it's one of the total sdk maintainers.

Author: It's explained below

> A voting manager, a member of the community entitled to manage the vote. The voting manager MUST be part of people entitled to vote.

  all people entitled to vote abstaining.
* One of the following conditions had been met:
  * The new maintainer(s) already meet the requirements enlisted in [New maintainers section](#new-maintainers)
  * The new maintainer(s) performed the required PRs, Reviews and Code changes to meet the [New maintainers criteria](#new-maintainers) but they were unable to merge them due to inactivity of the current maintainer(s).

The new maintainer must show the good will to keep the project healthy and
community could ask for an eventual plan to improve the quality of the project.

Once a new maintainer is identified, the community can proceed with a vote
to handover the project using the
[Asynchronous voting process](#asynchronous-voting-process). If more than one

> new maintainer wants to handover the project, separate voting process must be

Reviewer: s/handover/maintain/

Reviewer: I noticed that you have "must" here, not "MUST". I'd remove the uppercase version of the RFC2119 keywords in this doc since it's not a spec.

> done for all of them, and both of them must follow the same voting criteria (and

Reviewer: s/both/all/

not the ones from [New maintainers section](#new-maintainers)).

The voting criteria are:

- 1 week to vote

> - 2/3 of the agreement to accept the vote

Reviewer: at least 2/3 of the votes cast agree to the proposal

- All _sdk maintainers_ are entitled to vote

### Archive a project

If a project is not following the criteria 1, 2, 4, 5, 6, 7, 8, the community MAY decides to archive the project. Prior to archiving, the community SHOULD first consider performing a search for a new maintainer to [handover the project](#handover-to-a-new-maintainergroup-of-maintainers).

If no new maintainer is found, the community can proceed to archive the project
using the [Asynchronous voting process](#asynchronous-voting-process).

The voting criteria are:

- 2 weeks to vote

> - 2/3 of the agreement to accept the vote

Reviewer: at least 2/3 of the votes cast agree to the proposal

- All _sdk maintainers_ are entitled to vote

## Asynchronous voting process

In order to ensure the democratic management of debates in the community, we
define an asynchronous voting process using GitHub issues.

The voting is defined by:

- The proposed community change to vote
- The set of people/groups entitled to vote
- A voting period
- The percentage of agreements required to accept the change
- A voting manager, a member of the community entitled to manage the vote. The voting manager MUST be part of people entitled to vote.

To start a vote, the voting manager opens an issue in the
[CloudEvents Spec](https://github.com/cloudevents/spec/issues) with:

- A binary question to be voted on

> - Tags of the entitled people/groups to vote (depending on the question)

Reviewer: Identifies the list of people who are entitled to vote

Author: rephrased a bit

>
> Entitled to vote:
>
> - @sdk-go-maintainers
>
> Voting ends: 24th June 2020 12:00 CEST
>
> Required agreement: 50% + 1

Then every person entitled can vote, up to the expiration date, commenting on
the issue with a `+1` to express agreement or `-1` to express disagreement.
Every comment except `+1` and `-1` won't be taken in account, because we expect
that the voting process comes only after a debate on the change, developed on
the community channels and Serverless WG/CloudEvents SDK meetings. In order to
ensure visibility to the voting, people should be remembered to vote during
Serverless WG meetings and community channels (Slack, ...).

At the end of the voting period, the voting manager will count the votes,
writing on the issue the results. Votes after the expiration date MUST not be
considered; To calculate the agreement percentage, the voting manager should
account the people who have expressed a vote, without taking in account the
abstentions.

If the vote passes, the voting manager should proceed with the change. If the vote
does not pass, then a new vote for the same change MUST NOT happen for the
next 6 months.

> This voting process MUST be used only to resolve the issues identified in this

Reviewer: I'm not sure if this was the intent, but are you saying that this voting process can NOT be used for other topics? I'm wondering if you meant to say:
This voting process must be used in the following cases, but can be used in others as well:

Author: My intent is to avoid people starting _abusing_ of this (heavy) process for things that doesn't really need it.
If we want to add in future use cases where we want to use this process, then it's fine, we just need to change this document and agree on the changes

## SDK Community

The community is organized as follows:

- Every SDK has its own
  [Github Team](https://github.com/orgs/cloudevents/teams), for example
  `sdk-go-maintainers` is the group of mantainers for
  [cloudevents/sdk-go project](https://github.com/cloudevents/sdk-go)
- The union of all the `sdk-*-maintainers` assembles the _sdk maintainers_ group

## New maintainers

TBD

## Ensuring projects health

In order to ensure the health of the SDK projects, we define the following
actions. These are crucial to show the commitment of the community to develop
and support these libraries, which are an important part of the CloudEvents
ecosystem.

We define an SDK project _healthy_ if:

1. It works with the latest version of the programming language

> 2. It supports the latest versions of the integrated libraries/frameworks

Author: > Does this mean "dependencies are kept up to date"? If so, why not just say that?

That's not necessary true, in ecosystems like Java libraries could be bytecode backward compatibility. For example, if sdk-java depends on library-x v1.0, it could work with v1.1 or even with v2.0 without touching sdk-java.

> 2. It supports the latest versions of the integrated libraries/frameworks

Reviewer 2: So I guess I'm just not clear on what this means.

> 2. It supports the latest versions of the integrated libraries/frameworks

Author: Not sure how i could rephrase it honestly :man_shrugging:

> 3. It receives security patches regularly

Reviewer 2: ```suggestion
3. Any security patches received are applied and the library is published regularly
```
This phrasing is a little odd. What if there are no security issues with the library? Slightly modified to address that.

Author: If there are no security issues and all the other criteria are respected, then there is no need to publish regularly.
Your phrasing seems like saying "even if you meet 1, 2 and there is no need of security patches, library should be published regularly" which doesn't make sense to me honestly. If the library works, it's not buggy, it's secure and new 3rd party integrations can interact with it, there is no need to publish it.

Reviewer 2: OK, how about this?

```
3. Any security patches received are applied and published within one month of receipt
```

Author: but then it seems similar to criteria 8... Maybe should I remove one of the two?

4. It supports the last N-1 major versions of CloudEvents spec, as defined in
   [Contribution Acceptance](../SDK.md#contribution-acceptance)
5. Issues and PRs are triaged (labeled, commented, reviewed, etc) regularly

We define a project `cloudevents/sdk-x` _not actively maintained_ if:

6. No commits on master from `sdk-x-maintainers` since 4 months
7. Issues and/or PRs are not being triaged from `sdk-x-mantainers` since 2
   months
8. Security patches are not being **released** from `sdk-x-mantainers` since 1
   months from CVE disclosures

> To prevent projects _not actively maintained_, we define different actions the
> community might take:

Reviewer 2: ```suggestion
To prevent the project from becoming _not actively maintained_, the community MAY take the following actions:
```

- Temporary security patches delivered by an _sdk maintainer_ not part of

>   `sdk-x-maintainers` group

Reviewer 2: This sentence is a bit confusing. Two things stand out. First, what is an sdk maintainer? Is this someone who is in the sdk maintainers group but who is not a maintainer on X repository? Second, if so, how does this remedy the situation where a security patch is not released? Do the sdk maintainers (union of all sdk-x-maintainers) each have publish credentials for all SDKs? That seems insecure.

Author: I tried to explain it in the first section of the document the difference between _sdk maintainer_ and `sdk-x-maintainer`, how could i reword this sentence?

About having the rights to publish release, we had a discussion a while ago with @n3wscott which proposed to automate it, in order to avoid a single person to hold rights to publish packages. The part of this document about security patches kinda "assumes" that we have such automation in place

- Handover of the project to a new maintainer
- Archive the project

### Security patches

If a project `cloudevents/sdk-x` meets _healthy_ criteria 1, 2, 4, but the
`sdk-x-maintainers` group is not actively performing security patches as defined
in criteria 8, the community could decide to entitle one or more members from
the larger _sdk maintainers_ group to perform security patches and release new
minor versions of the project.

> Because the people entitled to perform security patches needs write rights on
> the repo, the community MUST temporary assign him/them the rights to perform it.
> After the release of the patch, these rights MUST be removed from the member(s)
> who took care of it.

Reviewer 2: ```suggestion
Because the people entitled to perform security patches need write permissions on the repository,
the community MUST temporarily grant these permissions.
After the release of the patch, these permissions MUST be revoked.
```

> Because of the urgency nature of security patches, an _informal_ vote during the

Reviewer 2: ```suggestion
Because of the urgent nature of security patches, an _informal_ vote during the
```

CloudEvents SDK meeting is enough to assign the security patch to the _sdk

> maintainer(s)_ which volunteers to take care of it.

Reviewer 2: ```suggestion
maintainer(s)_ who volunteers to take care of it.
```

NOTE: this process is valid only for security patches, not for bugs or
enhancements.

### Handover to a new maintainer/group of maintainers

> It may happen that, if a project is not following the criteria 1, 2, 4, 5, 6, 7,
> 8, the community might decide to handover the project to a new maintainer/group
> of maintainers. The community can perform the handover to a new maintainer
> **iff** a voting process to add such maintainer was **already performed** with
> 100% of entitled people abstained and:
> 
> - Or the new maintainer(s) already meet the requirements enlisted in
>   [New maintainers section](#new-maintainers),
> - Or, because of the inactivity of the project due to criteria 7, the new
>   maintainer(s) performed the required PRs/Reviews/Code changes to meet the
>   [New maintainers criteria](#new-maintainers) but he/they wasn't/weren't able
>   to merge them.

Reviewer 2: ```suggestion
If a project is not meeting the criteria 1, 2, 4, 5, 6, 7, 8, the community
MAY decide to handover the project to a new maintainer/group of
maintainers. The community can perform the handover to a new maintainer
under the following conditions.

* A vote to add the new maintainer MUST have **already taken place** with
  all existing repository maintainers abstaining, AND one of the following
  conditions has been met. 
* The new maintainer(s) already meet the requirements defined in
  [New maintainers section](#new-maintainers)
* The new maintainer(s) performed the required PRs, Reviews and Code
  changes to meet the [New maintainers criteria](#new-maintainers) but
  they were unable to merge them due to inaction by the current maintainer(s).
```

This section is a little confusing to me. Making suggestions here to clear it up, but it's confusing enough that I'm not sure this is actually what you meant.

Author: Took inspiration from your reword and reworded again, look at it now :smile:

The new maintainer must show the good will to keep the project healthy and
community could ask for an eventual plan to improve the quality of the project.

> As soon as the new maintainer is identified, the community can proceed to vote
> to handover the project using the

Reviewer 2: ```suggestion
Once a new maintainer is identified, the community can proceed with a vote
to handover the project using the
```

[Asynchronous voting process](#asynchronous-voting-process). If more than one
new maintainer wants to handover the project, separate voting process must be
done for all of them, and both of them must follow the same voting criteria (and

> not the ones from [New maintainers section](#new-maintainers)).

Reviewer 2: Not sure I understand the intent here. Is this addressing a situation where there are multiple community contributors to an unmaintained repository who each want to have the project handed over to them? This opens up larger questions around who "owns" the repository when there are multiple maintainers. I think this question needs to be addressed as well somewhere.

Author: That's exactly what i wanted to address. My point here is: if we have 2 people that wants to handover the project at same time, both must go through the same voting process with same criteria

> The voting criteria are:

Reviewer 2: What happens if the vote is "no"?

Author: It's explained below. If no the project remains as is. The community could decide to take the path of archiving the project

- 1 week to vote
- 2/3 of the agreement to accept the vote
- All _sdk maintainers_ are entitled to vote

### Archive a project

> It may happen that, if a project is not following the criteria 1, 2, 4, 5, 6, 7,
> 8, the community might decide to archive the project. Prior to archiving the
> project, the community should consider first to search a new maintainer to
> [handover the project](#handover-to-a-new-maintainergroup-of-maintainers).

Reviewer 2: ```suggestion
If a project is not following the criteria 1, 2, 4, 5, 6, 7, 8, the community
MAY decide to archive the project. Prior to archiving, the community
SHOULD first consider performing a search for a new maintainer to
[handover the project](#handover-to-a-new-maintainergroup-of-maintainers).
```

If no new maintainer is found, the community can proceed to archive the project
using the [Asynchronous voting process](#asynchronous-voting-process).

> The voting criteria are:

Reviewer 2: What happens if the vote is "no"?

Author: If no the project remains as is. I think the argument of somebody who might vote "no" could be somebody which thinks the project is already valuable, so he doesn't think it's a good idea to archive it

- 2 weeks to vote
- 2/3 of the agreement to accept the vote
- All _sdk maintainers_ are entitled to vote

## Asynchronous voting process

In order to ensure the democratic management of debates in the community, we
define an asynchronous voting process using GitHub issues.

The voting is defined by:

- The proposed community change to vote
- The set of people/groups entitled to vote
- A voting period
- The percentage of agreements required to accept the change
- A member of the community entitled to manage the vote, called voting manager.
  The voting manager MUST also be part of people entitled to vote.

> To start a vote, the voting manager opens an issue in the

Reviewer 2: Who is the "voting manager"?

Author: explained in the row above, let me try to rephrase it

[CloudEvents Spec](https://github.com/cloudevents/spec/issues) with:

> - The binary question which explains the change
> - The tags to the entitled people/groups to vote (depending on the question)
> - The expiration date of the vote
> - The percentage of agreements to accept the change

Reviewer 2: ```suggestion
- A question to be voted on
- Tags of the entitled people/groups to vote (depending on the question)
- The expiration date of the vote
- The percentage of votes required to accept the change
```

Does this voting process only take into account binary questions? Here is a suggestion that allows for a more flexible voting mechanism.

Author: > Does this voting process only take into account binary questions?

I would prefer to keep the binary questions, they're easier to manage and don't leave room to "confusion" on the vote result. Also at least in this document I don't see any use case for non-binary questions, we could always evolve such process later when we need it

For example:

> Question: Do you want to add "Paolo Rossi" @paolorossi as _sdk-go-maintainer_?
>
> Entitled to vote:
>
> - @sdk-go-maintainers
>
> Voting ends: 24th June 2020 12:00 CEST
>
> Required agreement: 50% + 1

Then every person entitled can vote, up to the expiration date, commenting on
the issue with a `+1` to express agreement or `-1` to express disagreement.
Every comment except `+1` and `-1` won't be taken in account, because we expect
that the voting process comes only after a debate on the change, developed on
the community channels and Serverless WG/CloudEvents SDK meetings. In order to
ensure visibility to the voting, people should be remembered to vote during

> Serverless WG meetings and community channels (Slack, ...).

Reviewer 2: What about using something like [gh-polls](https://gh-polls.com/)? I find going through and counting up +1 comments to be tedious and error prone - especially when you start asking yourself, does :+1: on the issue count as a yes vote.

Author: > I find going through and counting up +1 comments to be tedious and error prone - especially when you start asking yourself, does +1 on the issue count as a yes vote.

+1 is +1, i don't think it leaves too much room for interpretations :smile: I would prefer to avoid including another tool, because if we want to adopt this solution in other repos then we need to configure the tool in other repos and so on. It doesn't seem to me it adds enough value and +1 seems pretty straightforward to me. For the beginning IMO sticking with +1/-1 it's ok

At the end of the voting period, the voting manager will count the votes,
writing on the issue the results. Votes after the expiration date MUST not be
considered; To calculate the agreement percentage, the voting manager should
account the people who have expressed a vote, without taking in account the

> abstentions.

Reviewer 2: This section is unclear to me.

Author: It's about taking the abstentions out of the final count of votes. Let's take as example:

* 8 people are entitled to vote
* Required 50% + 1 for the vote to pass
* 3 effectively votes, the others don't
* 2 `-1`, 1 `+1`

Then the vote does not pass, because less than 50%+1 voted `+1`

> If the vote pass, the voting manager should proceed with the change. If the vote

Reviewer 2: ```suggestion
If the vote passes, the voting manager should proceed with the change. If the vote
```

> does not pass, then a new voting for the same change can't be started for the

Reviewer 2: ```suggestion
does not pass, then a new vote for the same change MAY NOT happen for the
```

Author: For me it's an hard requirement, it's a MUST NOT happen

next 6 months.

> This voting process MUST be used only to resolve the debates underlined in this

Reviewer 2: ```suggestion
This voting process MUST be used only to resolve the issues identified in this
```
